### PR TITLE
Fix: Compare input against beta * threshold in softplus backwards

### LINKED ISSUE #55587: `Softplus` forward and backward discrepancy
## 🐛 Bug

`Softplus` operator backward implementation different from doc and forward implementation on applying `threshold`.

Documentation defines:
`For numerical stability the implementation reverts to the linear function when input×β>threshold` 
[https://pytorch.org/docs/stable/generated/torch.nn.Softplus.html](https://pytorch.org/docs/stable/generated/torch.nn.Softplus.html)

This is consistent with the behavior of `Softplus` forward kernels. However on the backward computation,  the `input * beta > threshold` condition becomes `output * beta> threshold`. As shown in below code snippet from 

`master @ b39eeb07ed2f60a27b60ad32d172b58d31554282`

```cpp

// wrapper from ATen/native/Activation.cpp
Tensor softplus_backward(
    const Tensor& grad_output,
    const Tensor& self,
    const Scalar& beta,
    const Scalar& threshold,
    const Tensor& output) {
  Tensor grad_input;
  auto iter = TensorIterator::binary_op(grad_input, grad_output, output);
  softplus_backward_stub(iter.device_type(), iter, beta, threshold);
  return iter.output();
}

//CPU impl from ATen/native/cpu/Activation.cpp
cpu_kernel_vec(
        iter,
        [beta, threshold](scalar_t a, scalar_t b) -> scalar_t {
          scalar_t z = std::exp(b * beta);
          return (b * beta) > threshold ? a : a * (z - scalar_t(1.)) / z;
        },
        [beta_vec, one_vec, threshold_vec](Vec a, Vec b) -> Vec {
          const Vec z = (b * beta_vec).exp();
          return Vec::blendv(a * (z - one_vec) / z, a, (b * beta_vec) > threshold_vec);
        }
    );

// CUDA impl from ATen/native/cuda/Activation.cu
 AT_DISPATCH_FLOATING_TYPES_AND2(at::ScalarType::Half, at::ScalarType::BFloat16, iter.dtype(), "softplus_backward_cuda", [&]() {
    auto beta = beta_.to<scalar_t>();
    auto threshold = threshold_.to<scalar_t>();
    gpu_kernel(iter, [beta, threshold]GPU_LAMBDA(scalar_t a, scalar_t b) -> scalar_t {
      scalar_t z = std::exp(b * beta);
      return (b * beta) > threshold ? a : a * (z - scalar_t(1.)) / z;
    });
  });

```

This makes backward of `Softplus` op semantically mismatched with forward. Even though with the default threshold of 20 it may not have observable effect, it can break `gradcheck` when the threshold is lowered. 

## To Reproduce

Thanks @zasdfgbnm for the repro:

```cpp
import torch
model = torch.nn.Softplus(threshold=1).double()
input = torch.tensor(0.9, dtype=torch.double, requires_grad=True)
output = model(input)
print(input.item(), output.item())
torch.autograd.gradcheck(model, input)
```

`Threshold` is really a method of improving numerical stability and current implementation shouldn't have problem with high `threshold` values (default 20 should be ok). It might make sense not to expose `threshold` as a parameter, as it is not a hyper-parameter to be tuned for sure. Updating all the `softplus_backward` to use the input to compute the gradients sounds like the the most thorough way. 

This issue is intended to be used for some quick discussions. @zasdfgbnm , @mruberry 

cc @ezyang @gchanan @zou3519 @bdhirsh @jbschlosser @anjali411 @albanD @gqchen @pearu @nikitaved @soulitzer @mruberry

### PR DESCRIPTION
**Pull request by jbschlosser**
Fixes #55587

The fix converts the binary `TensorIterator` used by softplus backwards to a ternary one, adding in the original input for comparison against `beta * threshold`.

### CODE REVIEW

**Review by albanD — APPROVED** [review]
The fix looks good to me.
I don't know if there is a better to do that TensorIterator though... cc @ezyang ?

File: aten/src/ATen/native/Activation.cpp [hunk 1]
```diff
@@ -368,7 +378,17 @@ Tensor softplus_backward(
     const Scalar& threshold,
     const Tensor& output) {
   Tensor grad_input;
-  auto iter = TensorIterator::binary_op(grad_input, grad_output, output);
+  auto iter = TensorIteratorConfig()
```
Comment (bdhirsh) [inline aten/src/ATen/native/Activation.cpp:381]: I guess this is the first "ternary op" that uses TensorIterator?

I personally think it would be a little more maintainable just to add a new `TensorIterator::ternary_op()`, since that doesn't add too much cognitive load (we already have 5+ variants of that builder function). And that makes it easy to find all the call-sites of ternary ops, if/when more get added later.

**Comment by jbschlosser** [comment]
Looks like we need a change in XLA as well. cc @JackCaoG 
https://github.com/pytorch/xla/blob/f41a53458d12b19027f02e5ec34f51d187a63368/torch_xla/csrc/tensor_ops.cpp#L182-L190

Should be something like:
```python
XLATensor SoftplusBackward(const XLATensor& grad_output, const XLATensor& input,
                           const at::Scalar& beta, const at::Scalar& threshold,
                           const XLATensor& output) {
  XLATensor scaled_input = XLATensor::mul(input, beta);
  XLATensor scaled_output = XLATensor::mul(output, beta);
  XLATensor z = XLATensor::exp(scaled_output);
  return XLATensor::where(
      XLATensor::gt(scaled_input, threshold), grad_output,
      XLATensor::mul(grad_output, XLATensor::div(XLATensor::sub(z, 1, 1), z)));
}
```

File: aten/src/ATen/native/Activation.cpp (hunk 1 above)
Comment (jbschlosser) [reply aten/src/ATen/native/Activation.cpp:381]: Cool, sounds good. I'll add `ternary_op()`.

**Comment by JackCaoG** [comment]
@jbschlosser Thanks for the heads up, I can submit a corresponding pt/xla pr.

File: aten/src/ATen/TensorIterator.cpp [hunk 2]
```diff
@@ -743,6 +743,35 @@ void TensorIteratorBase::select_all_keeping_dim(int start_dim, IntArrayRef indic
   }
 }
 
+void TensorIteratorBase::build_ternary_float_op(
```
Comment (ezyang) [inline aten/src/ATen/TensorIterator.cpp:746]: cc @wenleix

**Comment by ezyang** [comment]
The TI new config seems fine.

**Comment by wenleix** [comment]
Added `ternary_op` in TI LGTM.

File: aten/src/ATen/native/cuda/Activation.cu [hunk 3]
```diff
@@ -297,9 +297,9 @@ void softplus_backward_kernel(TensorIterator& iter, const Scalar& beta_, const S
   AT_DISPATCH_FLOATING_TYPES_AND2(at::ScalarType::Half, at::ScalarType::BFloat16, iter.dtype(), "softplus_backward_cuda", [&]() {
     auto beta = beta_.to<scalar_t>();
     auto threshold = threshold_.to<scalar_t>();
-    gpu_kernel(iter, [beta, threshold]GPU_LAMBDA(scalar_t a, scalar_t b) -> scalar_t {
+    gpu_kernel(iter, [beta, threshold]GPU_LAMBDA(scalar_t a, scalar_t b, scalar_t c) -> scalar_t {
```
Comment (shmsong) [inline aten/src/ATen/native/cuda/Activation.cu:300]: Just a quick thought, a possible way around having to introduce `ternary_op` is to pass `in` instead of `out` as b, and then the below line becomes:
```cpp
 return (b * beta) > threshold ? a : a * z / (z + scalar_t(1.)) ;
```

Comment (jbschlosser) [reply aten/src/ATen/native/cuda/Activation.cu:300]: @shmsong I thought about doing this originally, but it gives incorrect results (note that `z` is computed from `b`, so its value changes if `b` changes).

Both input and output need to be available to the backward pass computation- input is used for the threshold and output is used for the value itself.

Comment (shmsong) [reply aten/src/ATen/native/cuda/Activation.cu:300]: Yes `z` will be different and the part using `z` also needs to be updated: (I should have highlighted this part in the original comment)
```
a * (z - scalar_t(1.)) / z
``` 
-> 
```
a * z / (z + scalar_t(1.))
```
There is a useful identity on softplus:
```
 exp(beta*out) == exp(beta*in) + 1
```
[https://pytorch.org/docs/stable/generated/torch.nn.Softplus.html](https://pytorch.org/docs/stable/generated/torch.nn.Softplus.html)

This should give correct result if I understand it correctly.

Comment (jbschlosser) [reply aten/src/ATen/native/cuda/Activation.cu:300]: Ah yes, thanks for clarifying. Looks right to me and avoids the addition of `ternary_op`.
